### PR TITLE
Allow quoted identifiers

### PR DESCRIPTION
**Pull request by rossberg**
Address #21.

Allow arbitrary non-empty names (well-formed unicode string literals) as identifiers.

Extend spec, interpreter, and test suite.

Note: This PR does not per se remove the explicit @name annotations from the proposal, because they are not equivalent: they express the actual presence of a corresponding custom section, whereas the use of symbolic names doesn't, even though they can be utilised by a tool to _construct_ such a custom section (as before). Also, annotations are more expressive, e.g., can provide alternative names to the same binder, i.e., quoted identifiers are not sufficient to convert all forms of name section into text.

### CODE REVIEW

**Comment by tlively** [comment]
The name annotation might technically be more expressive, but is there any demand for it? We shouldn't spec something that nobody implements or uses.

**Comment by rossberg** [comment]
Well, the idea underlying this proposal is that every custom section should have a text equivalent.

File: document/core/text/values.rst [hunk 1]
```diff
@@ -211,12 +211,13 @@ Identifiers
...
 :ref:`Indices <syntax-index>` can be given in both numeric and symbolic form.
-Symbolic *identifiers* that stand in lieu of indices start with :math:`\text{\$}`, followed by any sequence of printable |ASCII|_ characters that does not contain a space, quotation mark, comma, semicolon, or bracket.
+Symbolic *identifiers* that stand in lieu of indices start with :math:`\text{\$}`, followed by eiter a sequence of printable |ASCII|_ characters that does not contain a space, quotation mark, comma, semicolon, or bracket, or by a quoted :ref:`name <text-name>`.
 
 .. math::
    \begin{array}{llclll@{\qquad}l}
    \production{identifier} & \Tid &::=&
-     \text{\$}~\Tidchar^+ \\
+     \text{\$}~c^\ast{:}\Tidchar^+ &\Rightarrow& c^\ast \\ &&|&
+     \text{\$}~c^\ast{:}\Tname &\Rightarrow& c^\ast & (\iff |c^\ast| > 0) \\
```
Comment (tlively) [inline document/core/text/values.rst:220]: Why disallow the empty name? `$""` seems like it should be a valid identifier.

File: document/core/text/lexical.rst [hunk 2]
```diff
@@ -133,7 +133,7 @@ The *look-ahead* restrictions on the productions for |Tblockchar| disambiguate t
 Annotations
 ~~~~~~~~~~~
 
-An *annotation* is a bracketed token sequence headed by an *annotation id* of the form :math:`\T{@id}`.
+An *annotation* is a bracketed token sequence headed by an *annotation id* of the form :math:`\text{@id}` or :math:`\text{@"..."}`.
```
Comment (tlively) [inline document/core/text/lexical.rst:138]: The rendering of `\text{@"..."}` is pretty ugly. Can we use `\ldots` in there to make it prettier?

**Comment by tlively** [comment]
> Well, the idea underlying this proposal is that every custom section should have a text equivalent.

The syntax for arbitrary custom sections means that this goal is met without the `@name` annotation, albeit not in a very useful way. I would still like to see evidence that anyone wants to use the `@name` annotations. @alexcrichton, you were concerned about reduced expressivity without the `@name` annotations in [this comment](https://github.com/WebAssembly/annotations/issues/21#issuecomment-1757097765). What are you current thoughts on the usefulness of the `@name` annotations?

**Comment by alexcrichton** [comment]
In the abstract I do not personally have a use case for `@name` insofar as all practical wasm modules I've worked with all have a `name` section where names are unique and fit the existing identifier grammer. I've only used `@name` in testing/fuzzing/etc, never for a real-world module.

My comment there though was also specifically addressing the point that if the purpose of `@name` is to have a text format representation for any valid `name` section then somehow duplicate names need to be handled as well. Removing `@name`, which this PR is not proposing, would mean that such a `name` section could not be represented in the text format. I don't know if this is a big issue though. A `name` section which names function 10 in a module of 8 functions also can't be represented in the text format.

**Comment by rossberg** [comment]
> A `name` section which names function 10 in a module of 8 functions also can't be represented in the text format.

I'm afraid that's apples vs potatoes: the existing spec of the name section already requires referenced indices to exist, but at the same time explicitly allows multiple names for a given index. The custom section validator implemented for the reference interpreter as part of this proposal even checks it accordingly. ;)

**Pull request merged**